pyqg/xarray_output.py

```diff
@@ -0,0 +1,152 @@
+        'tc': {'long_name': 'model timestep', 'units': 'seconds',},
+        'dt': {'long_name': 'numerical timestep', 'units': 'seconds',},
+        'L': {'long_name': 'domain length in x direction', 'units': 'meters',},
+        'W': {'long_name': 'domain length in y direction', 'units': 'meters',},
+        'filterfac': {'long_name': 'amplitude of spectral spherical filter', 'units': '',},
+        'twrite': {'long_name': 'interval for cfl writeout', 'units': 'number of timesteps',},
+        'tmax': {'long_name': 'total time of integration', 'units': 'seconds',},
+        'tavestart': {'long_name': 'start time for averaging', 'units': 'seconds',},
+        'tsnapstart': {'long_name': 'start time for snapshot writeout', 'units': 'seconds'},
+        'taveint': {'long_name': 'time interval for accumulation of diagnostic averages', 'units': 'seconds'},
+        'tsnapint': {'long_name': 'time interval for snapshots', 'units': 'seconds',},
+        'ntd': {'long_name': 'number of threads used', 'units': 'none',},
+        'pmodes': {'long_name': 'vertical pressure modes', 'units': 'none',},
+        'radii': {'long_name': 'deformation radii', 'units': 'meters',},
+    }
+
+    # dict for dataset attributes
+    ds_attr_database = {
+        'title': 'pyqg: Python Quasigeostrophic Model',
+        'institution': '',
+        'source': ('version: {}'.format(pyqg.__version__)),
+        'history': '',
+        'references': 'https://pyqg.readthedocs.io/en/latest/index.html',
+        'comment': '', 
+    }
+
+
+    # Create list of variables
+    variables = {}
+    for vname in dim_database:
+        data = getattr(m,vname)
+        if 'time' in dim_database[vname]:
+            variables[vname] = (dim_database[vname],data[np.newaxis,...])
+        else:
+            variables[vname] = (dim_database[vname],data)
+
+    # Create list of coordinates
+    coordinates = {}
+    cname_to_1D = ['time','z','x','y','l','k']
+    coords_1D = [(np.array([m.t])), (np.arange(1,m.nz+1)), (m.x[0,:]), (m.y[:,0]), (m.l[:,0]), (m.k[0,:])] 
+    for cname in coord_database:
+        if cname in cname_to_1D:
+            index = cname_to_1D.index(cname)
+            coordinates[cname] = coords_1D[index]
+        else:
+            try:
+                data = getattr(m,cname)
+                coordinates[cname] = (coord_database[cname],data )
+            except:
+                pass
+
+    # Define dataset
+    ds = xr.Dataset(variables,
+                    coords=coordinates,
+                    attrs=ds_attr_database)
+
+    # Assign attributes to coordinates
+    for caname in coord_attr_database:
+        if caname in ds.coords:
+            ds.coords[caname].attrs = coord_attr_database[caname]
+
+    # Assign attributes to variables
+    for vaname in var_attr_database:
+        if vaname in ds.data_vars:
+            ds.data_vars[vaname].attrs = var_attr_database[vaname]
```

Comment:
Let's move these loops into the loops above so we only have to do each loop (vars and coords) once.

Reply:
I've combined the loops for vname and cname [here](https://github.com/pyqg/pyqg/pull/230/commits/274fb879275bcd373a6ef230eeddcc1f38664c3d) (see code snippet below). However I'm now essentially filling in the Dataset within the loops and I'm not sure if this is the best way of going about it. 

```python 
def model_to_dataset(m):

ds = xr.Dataset()
    
    # Create list of variables
    variables = {}
    for vname in dim_database:
        if hasattr(m,vname):
            data = getattr(m,vname, None)
            if 'time' in dim_database[vname]:
                variables[vname] = (dim_database[vname],data[np.newaxis,...])
                ds = xr.merge([ds, ds.assign(variables)])
                ds.data_vars[vname].attrs = var_attr_database[vname]
            else:
                variables[vname] = (dim_database[vname],data)
                ds = xr.merge([ds, ds.assign(variables)])
                ds.data_vars[vname].attrs = var_attr_database[vname]

    # Create list of coordinates
    coordinates = {}
    for cname in coord_database:
        if cname in transformations:
            coordinates[cname] = transformations[cname](m)
            ds = xr.merge([ds, ds.assign_coords(coordinates)])
            ds.coords[cname].attrs = coord_attr_database[cname]
        else:
            if hasattr(m,cname):
                data = getattr(m, cname)
                coordinates[cname] = (coord_database[cname],data)
                ds = xr.merge([ds, ds.assign_coords(coordinates)])
                ds.coords[cname].attrs = coord_attr_database[cname

    ds.attrs = ds_attr_database 

    return ds
```

Reply:
I would do it like this

```python
for vname in dim_database:
    # do stuff
    variables[vname] = (dims, data, attrs)
for cname in coord_database:
    # do stuff
    coords[cname] = (dims, data, attrs
ds = xr.Dataset(variables, coords=coords, global_attrs)
```

Reply:
Made this change, thanks!